File: Data Structures & Algorithms/binary-tree-right-side-view/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        lo = []
        
        def dfs(root, depth):
            if not root:
                return None
            if(len(lo) == depth):
                lo.append([])
            lo[depth].append(root.val)
            dfs(root.left, depth+1)
            dfs(root.right, depth+1)
        
        dfs(root, 0)
        print(lo)
        res = [l[-1] for l in lo]
        print(res)
        return res

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int result = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        helper(root);
        return result;
    }

    int helper(TreeNode root) {
        if(root == null) return 0;
        int leftMax = Math.max(0, helper(root.left));
        int rightMax = Math.max(0, helper(root.right));
        int sum = root.val + leftMax + rightMax;
        result = Math.max(result, sum);
        return root.val + Math.max(leftMax, rightMax);
    }
}
